crates/noirc_evaluator/src/ssa/acir_gen.rs

```diff
@@ -625,81 +637,100 @@ pub fn evaluate_cmp(
     }
 }
 
-//Decomposition into a list of unsigned integers based on a base2 radix (bits - u1, bytes - u8, etc)
-pub fn to_base2_decomposition(
-    pow: u32,
+const fn num_bits<T>() -> usize {
+    std::mem::size_of::<T>() * 8
+}
+pub fn bit_size_u32(a: u32) -> u32 where {
+    num_bits::<u32>() as u32 - a.leading_zeros()
+}
+
+pub fn bit_size_u128(a: u128) -> u32 where {
+    num_bits::<u128>() as u32 - a.leading_zeros()
+}
+
+//Decomposition into b-base: \sum ai b^i, where 0<=ai<b
+// radix: the base, (it is a constant, not a witness)
+// num_limbs: the number of elements in the decomposition
+// output: (the elements of the decomposition as witness, the sum expression)
+pub fn to_radix(
+    radix: u32,
     num_limbs: u32,
     evaluator: &mut Evaluator,
 ) -> (Vec<Witness>, Expression) {
     let mut digits = Expression::default();
-    let mut two_pow = FieldElement::one();
-    let base: i32 = 2;
-    let radix_base = base.pow(pow);
-    let shift = FieldElement::from(radix_base as i128);
+    let mut radix_pow = FieldElement::one();
+
+    let shift = FieldElement::from(radix as i128);
     let mut result = Vec::new();
+    let bit_size = bit_size_u32(radix);
     for _ in 0..num_limbs {
-        let radix_witness = evaluator.add_witness_to_cs();
-        result.push(radix_witness);
-        let radix_expr = from_witness(radix_witness);
-        digits = add(&digits, two_pow, &radix_expr);
-        two_pow = two_pow.mul(shift);
-
-        range_constraint(radix_witness, pow, evaluator).unwrap();
+        let limb_witness = evaluator.add_witness_to_cs();
+        result.push(limb_witness);
+        let limb_expr = from_witness(limb_witness);
+        digits = add(&digits, radix_pow, &limb_expr);
+        radix_pow = radix_pow.mul(shift);
+
+        if 1_u128 << (bit_size - 1) != radix as u128 {
+            range_constraint(limb_witness, bit_size, evaluator).unwrap();
+        }
+        bound_constraint_with_offset(
+            &from_witness(limb_witness),
+            &Expression::from_field(shift),
+            &Expression::one(),
+            bit_size,
+            evaluator,
+        );
     }
 
     (result, digits)
 }
 
-//Performs byte decomposition
-pub fn to_bytes(lhs: &InternalVar, byte_size: u32, evaluator: &mut Evaluator) -> Vec<Witness> {
-    assert!(byte_size < FieldElement::max_num_bytes());
-    let (result, bytes) = to_base2_decomposition(8, byte_size, evaluator);
-    evaluator.gates.push(Gate::Directive(Directive::ToBytes {
+//decompose lhs onto radix-base with limb_size limbs
+pub fn to_radix_base(
+    lhs: &InternalVar,
+    radix: u32,
+    limb_size: u32,
+    evaluator: &mut Evaluator,
+) -> Vec<Witness> {
+    // ensure there is no overflow
+    let mut max = BigUint::from(radix);
+    max = max.pow(limb_size) - BigUint::one();
+    assert!(max < FieldElement::modulus());
+
+    let (result, bytes) = to_radix(radix, limb_size, evaluator);
```

Comment:
to_radix always return bytes, this is because bigint has this limitation. We can easily remove this limitation in future if it is required to support bigger radix size.